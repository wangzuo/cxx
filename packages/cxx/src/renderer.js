import React from 'react';
import { renderToString } from 'react-dom/server';
import { createOperationSelector, getOperation } from 'relay-runtime';
import html from './html';
import normalizeHref from './normalize-href';
import createMatcher from './matcher';
import QueryRenderer from './query-renderer';
import createServerEnvironment from './createServerEnvironment';

export default (routes = {}, schema) => {
  const matcher = createMatcher(routes);

  return async (href, options = {}) => {
    const context = options.context || null;
    const assets = options.assets || {};
    const location = normalizeHref(href);
    const matchLocation = matcher(location);

    if (!matchLocation) {
      return '404 not found';
    }

    const { page } = matchLocation;
    if (!page.query) {
      const markup = renderToString(<page.default />);
      return html(markup, {}, assets);
    }

    const variables = page.variables ? page.variables(matchLocation) : {};
    const operation = createOperationSelector(
      getOperation(page.query),
      variables
    );
    const environment = createServerEnvironment(schema, context);
    environment.retain(operation.root);
    await new Promise((resolve, reject) => {
      environment.sendQuery({
        operation,
        onCompleted: resolve,
        onError: reject
      });
    });

    const markup = renderToString(
      <QueryRenderer
        environment={environment}
        query={page.query}
        variables={variables}
        render={result => React.createElement(page.default, result)}
      />
    );

    const initial = environment.getStore().getSource().toJSON();

    return html(markup, initial, assets);
  };
};
