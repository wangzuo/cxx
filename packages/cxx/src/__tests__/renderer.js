import React from 'react';
import { GraphQLSchema, GraphQLObjectType, GraphQLString } from 'graphql';
import renderer from '../renderer';
import Head from '../head';

Head.canUseDOM = false;

const schema = new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'Query',
    fields: {
      hello: {
        type: GraphQLString,
        resolve: () => 'Hello'
      }
    }
  })
});

test('render without schema', async () => {
  const page = {
    default: () => <div>test</div>
  };
  const routes = {
    '/': page
  };

  const html = await renderer(routes)('/');
  expect(html).toMatchSnapshot();
});

test('render with head', async () => {
  const page = {
    default: () =>
      <div>
        <Head>
          <title>test</title>
        </Head>
        test
      </div>
  };
  const routes = {
    '/': page
  };

  const html = await renderer(routes)('/');
  expect(html).toMatchSnapshot();
});

test('default 404', async () => {
  const html = await renderer()('/');
  expect(html).toMatchSnapshot();
});
