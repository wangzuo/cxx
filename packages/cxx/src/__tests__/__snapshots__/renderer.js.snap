// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default 404 1`] = `"404 not found"`;

exports[`render with head 1`] = `
"<!doctype html>
  <html>
    <head>
      <title data-react-helmet=\\"true\\">test</title>
      
      
      
      </head>
    <body >
      <div id=\\"__cxx__\\"><div data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"1171596156\\"><!-- react-empty: 2 --><!-- react-text: 3 -->test<!-- /react-text --></div></div>
      <script>var __DATA__= {};</script>
      
      
      
      
    </body>
  </html>"
`;

exports[`render without schema 1`] = `
"<!doctype html>
  <html>
    <head>
      <title data-react-helmet=\\"true\\"></title>
      
      
      
      </head>
    <body >
      <div id=\\"__cxx__\\"><div data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"-1145761510\\">test</div></div>
      <script>var __DATA__= {};</script>
      
      
      
      
    </body>
  </html>"
`;
